modules/packages/OrderedSet.chpl

```diff
@@ -0,0 +1,484 @@
+      this.eltType = eltType;
+      this.parSafe = parSafe;
+      this.implType = implType;
+
+      this.instance = getInstanceFromEnumVal(implType, eltType, parSafe, comparator); 
+    }
+
+    /*
+      Initialize this orderedSet with a unique copy of each element contained in
+      `iterable`. If an element from `iterable` is already contained in this
+      orderedSet, it will not be added again. The formal `iterable` must be a type
+      with an iterator named "these" defined for it.
+
+      :arg iterable: A collection of elements to add to this orderedSet.
+      :arg parSafe: If `true`, this orderedSet will use parallel safe operations.
+      :type parSafe: bool
+      :arg comparator: The comparator used to compare elements.
+    */
+    proc init(type eltType, iterable, param parSafe=false,
+              comparator: record = defaultComparator,
+              param implType: setImpl = defaultImpl)
+    where canResolveMethod(iterable, "these") lifetime this < iterable {
+      this.eltType = eltType;
+      this.parSafe = parSafe;
+      this.implType = implType;
+
+      this.instance = getInstanceFromEnumVal(implType, eltType, iterable,
+                                            parSafe, comparator); 
+    }
+
+    /*
+      Initialize this orderedSet with a copy of each of the elements contained in
+      the orderedSet `other`. This orderedSet will inherit the `parSafe` value of 
+      the orderedSet `other`.
+
+      :arg other: An orderedSet to initialize this orderedSet with.
+    */
+    proc init=(const ref other: orderedSet(?t)) lifetime this < other {
+      this.eltType = t;
+      this.parSafe = other.parSafe;
+      this.instance = getInstanceFromEnumVal(this.implType, this.eltType,
+                                            this.parSafe,
+                                            other.instance.comparator); 
+
+      this.complete();
+
+
+      if !isCopyableType(eltType) then
+        compilerError('Cannot initialize ' + this.type:string + ' from ' +
+                      other.type:string + ' because element type ' +
+                      eltType:string + ' is not copyable');
+
+      for elem in other do instance._add(elem);
+    }
+
+    /*
+      Write the contents of this orderedSet to a channel.
+
+      :arg ch: A channel to write to.
+    */
+    proc const writeThis(ch: channel) throws {
+      instance.writeThis(ch);
+    }
+  }
+
+  /*
+    NOTE: Operators are borrowed from Set.chpl
+  */
+
+  /*
+    Clear the contents of this orderedSet, then extend this now empty orderedSet 
+    with the elements contained in another orderedSet.
+
+    .. warning::
+
+      This will invalidate any references to elements previously contained in
+      `lhs`.
+
+    :arg lhs: The orderedSet to assign to.
+    :arg rhs: The orderedSet to assign from. 
+  */
+  proc =(ref lhs: orderedSet(?t), rhs: orderedSet(?r)) {
+    lhs.clear();
+    for x in rhs {
+      lhs.add(x);
+    }
```

Comment:
:) Right, I wasn't thinking too much beyond "can we avoid doing n separate additions". And to be clear, it is OK to not do that at first and leave it as a future work. Especially so if it is not clear how to achieve that implementation-wise.

Reply:
> and could be improved to O(n)

Is that possible without adding new interface or touching the internal?

Reply:
Yeah maybe not. I think it can be a good future step for this module. We can add some support functions to the internal implementation and use it in the interface if the internal implementation supports it.